### PR TITLE
Simplified strict typechecking

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
## Description of changes

https://github.com/cedar-policy/rfcs/pull/19

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy-validator/src/typecheck.rs [hunk 1]
```diff
@@ -2483,10 +2007,21 @@ impl<'a> Typechecker<'a> {
...
             Some(actual_ty) => {
-                if !expected
-                    .iter()
-                    .any(|expected_ty| Type::is_subtype(self.schema, actual_ty, expected_ty))
-                {
+                if !expected.iter().any(|expected_ty| {
+                    // This check uses `ValidationMode::Permissive` even in
+                    // strict typechecking because we use this function and
+                    // `expect_type` to require that an operand is a record type
+                    // or an entity type by expecting `AnyEntity` to `{}`. In
```
Comment (khieta) [inline cedar-policy-validator/src/typecheck.rs:2014]: > by expecting `AnyEntity` to `{}`

Not sure what this means

File: cedar-policy-validator/src/types.rs [hunk 2]
```diff
@@ -1075,14 +1154,53 @@ impl EntityRecordKind {
...
+                },
+                ActionEntity {
+                    name: action_type2,
+                    attrs: attrs2,
+                },
+            ) => {
+                if action_type1 == action_type2 {
+                    // Same record type. Ensure that they have the same
+                    // attributes.  Computing the LUB under strict mode ensures
+                    // the attributes are the same. This will no-op as along as
```
Comment (khieta) [inline cedar-policy-validator/src/types.rs:1170]: will no-op -> will be a no-op
as along as -> as long as

File: cedar-policy-validator/src/types.rs [hunk 3]
```diff
@@ -1075,14 +1154,53 @@ impl EntityRecordKind {
...
+                    // attributes don't have any attributes, letting us define
```
Comment (khieta) [inline cedar-policy-validator/src/types.rs:1171]: attributes (first occurrence) -> actions

File: cedar-policy-validator/src/types.rs [hunk 4]
```diff
@@ -1075,14 +1154,53 @@ impl EntityRecordKind {
...
+                if mode.is_strict() && lub0 != lub1 {
+                    None
+                } else {
+                    Some(Entity(lub0.least_upper_bound(lub1)))
+                }
+            }
 
-            (AnyEntity, AnyEntity)
-            | (AnyEntity, Entity(_))
+            (AnyEntity, AnyEntity) => Some(AnyEntity),
```
Comment (khieta) [inline cedar-policy-validator/src/types.rs:1192]: `AnyEntity` shouldn't be allowed for strict mode, right?

File: cedar-policy-validator/src/types.rs [hunk 5]
```diff
@@ -1125,12 +1250,20 @@ impl EntityRecordKind {
                 // `rk1`.  When `rk1` is closed, a subtype of `rk1` may not have
                 // any attributes that are not listed in `rk1`, so we apply
                 // depth subtyping only.
```
Comment (khieta) [inline cedar-policy-validator/src/types.rs:1252]: Update comment to reflect that you also use depth subtyping for strict mode

**Review by khieta — APPROVED** [review]
A few tiny comments. Most of my comments are on [cedar-spec#111](https://github.com/cedar-policy/cedar-spec/pull/111) because the spec is easier (for me) to read.

**Review by aaronjeline — COMMENTED** [review]
Seems reasonable explained, as long as DRT passes.

**Review by aaronjeline — APPROVED** [review]

**Pull request merged**